src/lib.rs

```diff
@@ -1,41 +1,278 @@
+    /// * `config`: Consists of `duration`, `start_price`, `tokens_goal`, `price_increase_step` and time_increase_step 
+    /// 
+    async fn start_ico(&mut self, config: IcoAction) {
+        check_input(&config);
+        assert_owner_message(&self.owner, "start_ico(): Not owner start ICO");
+        assert!(!self.ico_state.ico_started, "start_ico(): Second ICO start");
+
+        if let IcoAction::StartSale { duration, start_price, tokens_goal, price_increase_step, time_increase_step } = config {
+            self.start_price = start_price;
+            self.tokens_goal = tokens_goal;
+            self.price_increase_step = price_increase_step;
+            self.time_increase_step = time_increase_step;
+
+            transfer_tokens(&self.token_id, &self.owner, &exec::program_id(), self.tokens_goal).await;
+
+            self.ico_state.ico_started = true;
+            self.ico_state.duration = duration;
+            self.ico_state.start_time = exec::block_timestamp();
+
+            msg::reply(IcoEvent::SaleStarted {
+                duration,
+                start_price,
+                tokens_goal,
+                price_increase_step,
+                time_increase_step,
+            }, 0).unwrap();
+        }
+    }
+
+    /// Purchase of tokens
+    /// 
+    /// Requirements:
+    /// * `tokens_cnt` must be greater than zero
+    /// * ICO must be in progress (already started and not finished yet)
+    /// * `msg::value` must be greater than or equal to `price * tokens_cnt`
+    /// * At least `tokens_cnt` tokens available for sale
+    /// 
+    /// Arguments:
+    /// * `tokens_cnt`: amount of tokens to purchase
+    /// 
+    pub fn buy_tokens(&mut self, tokens_cnt: u128)  {
+        let time_now: u64 = exec::block_timestamp();
+
+        assert!(tokens_cnt != 0, "buy_tokens(): Can't buy zero tokens");
+        assert!(self.ico_state.start_time + self.ico_state.duration >= time_now, "buy_tokens(): Duration of the ICO has ended");
+        assert!(self.get_balance() != 0, "buy_tokens(): All tokens have been sold");
+        self.in_process("buy_tokens()");
+
+        assert!(tokens_cnt <= self.get_balance(), "buy_tokens(): Not enough tokens to sell");
+
+        let current_price = self.get_current_price(time_now);
+        let cost = tokens_cnt.checked_mul(current_price)
+            .unwrap_or_else(|| panic!("buy_tokens(): Overflowing multiplication: {} * {}", tokens_cnt, current_price));
+
+        let mut change = 0;
+        let amount_sent = msg::value();
+
+        assert!(amount_sent >= cost, "buy_tokens(): Wrong amount sent, expect {} get {}", cost, amount_sent);
+
+        if amount_sent > cost {
+            change = amount_sent - cost;
+            msg::send(msg::source(), "", change).unwrap();
+        }
+
+        self.token_holders
+            .entry(msg::source())
+            .and_modify(|balance| *balance += tokens_cnt)
+            .or_insert(tokens_cnt);
+
+        self.tokens_sold += tokens_cnt; 
```

Comment:
Not protected against overflow (in theory).

Reply:
It was discussed here that the check is unnecessary

https://github.com/gear-tech/apps/pull/52#discussion_r894488054

Reply:
Ah, ok.